### DISCUSSION TITLE
Binding Control event's like KeyDown to ViewModel?

### DISCUSSION THREAD

**Discussion by Shadowblitz16** [AvaloniaUI/Avalonia#16360] [Q&A]
How do I bind any event to my ViewModel? 
Right now I am trying to bind KeyDown on the 3rd part ZoomBorder control.
```xml
            <ScrollViewer Grid.Row="1" Margin="1">
                <pz:ZoomBorder KeyDown="{Binding CanvasKeyDown}" Stretch="Uniform" EnableConstrains="True" Focusable="True" MaxZoomX="4" MaxZoomY="4" MinZoomX="1" MinZoomY="1" VerticalAlignment="Stretch" HorizontalAlignment="Stretch">
                    <Canvas Background="Black" Width="256" Height="240">
                        <Rectangle Stroke="Magenta" StrokeThickness = "1" Width="256" Height="240"/>
                    </Canvas>
                </pz:ZoomBorder>
            </ScrollViewer>
```

```cs
public partial class MainViewModel : ObservableObject
{
    public void CanvasKeyDown(object? sender, KeyEventArgs? args)
    {
        if (sender is ZoomBorder panAndZoom)
        {
            if (args.Key == Key.Up) panAndZoom.PanDelta(0, -16);
            if (args.Key == Key.Down) panAndZoom.PanDelta(0, +16);
            if (args.Key == Key.Left) panAndZoom.PanDelta(-16, 0);
            if (args.Key == Key.Right) panAndZoom.PanDelta(+16, 0);
        }
    }
}
```

**Comment by Shadowblitz16** [comment]
I also tried this...
```cs
public partial class MainViewModel : ObservableObject
{
    public ICommand CanvasKeyDownCommand { get; }

    public MainViewModel()
    {
        CanvasKeyDownCommand = new RelayCommand<KeyEventArgs>(CanvasKeyDown);
    }
    
    private void CanvasKeyDown(KeyEventArgs? args)
    {
        if (args?.Source is ZoomBorder panAndZoom)
        {
            if (args.Key == Key.Up   ) panAndZoom.PanDelta(0, -16);
            if (args.Key == Key.Down ) panAndZoom.PanDelta(0, +16);
            if (args.Key == Key.Left ) panAndZoom.PanDelta(-16, 0);
            if (args.Key == Key.Right) panAndZoom.PanDelta(+16, 0);
        }
    }
}
```
```
0>MainView.axaml(43,32): Error AVLN:0004 Avalonia: Unable to find suitable setter or adder for property KeyDown of type Avalonia.Base:Avalonia.Input.InputElement for argument Avalonia.Markup.Xaml:Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindingExtension, available setter parameter lists are:
System.EventHandler`1<Avalonia.Input.KeyEventArgs> Line 43, position 32.
0>------- Finished building project: MarioKit.Editor. Succeeded: False. Errors: 1. Warnings: 0
```

**Comment by stevemonaco** [comment]
Events are meant to be handled in code-behind. Write the handler in code-behind and forward the call to the ViewModel.

Handling the event as-is on the ViewModel means you're introducing Avalonia types into the ViewModel which is something I strongly try to avoid.

**Reply by Shadowblitz16** [reply]
Wait I thought the point of mvvm was to separate ui and logic?

**Reply by stevemonaco** [reply]
You should think of logic in at least two categories: view logic and domain logic (or core problem-solving logic). MVVM separates those.

Handling UI input events is UI. Your ViewModel doesn't need to know that the user pressed F5, the ViewModel needs to be aware that it needs to refresh your browser page. So you write a `KeyDown` handler (or `KeyBinding`), the View handles the input event and calls the ViewModel to execute some command or method to do the work.

**Comment by workgroupengineering** [comment]
Using https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors

**Reply by Shadowblitz16** [reply]
> Using https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors

I don't know what this is, I looked in the examples folder and still don't understand.

**Reply by stevemonaco** [reply]
They mean `EventTriggerBehavior` along with `InvokeCommandAction` like https://github.com/AvaloniaUI/Avalonia.Xaml.Behaviors/blob/d62d2aa68e3943b26b8f33cc120f2e8cc707e038/samples/BehaviorsTestApplication/Views/Pages/InvokeCommandActionView.axaml#L24 shows, except with `KeyDown`.

I don't recommend this because it's: 1. incredibly verbose and 2. isn't trim-compatible, AFAIK. You aren't gaining much here by keeping everything in XAML.

**Comment by hlz2516** [comment]
Use [Observable.FromEventPattern](https://learn.microsoft.com/en-us/previous-versions/dotnet/reactive-extensions/hh211984(v=vs.103)).  
you can bind command like this:
```
   // in your view's Constructor
        InitializeComponent();
        this.WhenActivated((disposable) =>
        {
            Observable.FromEventPattern<KeyEventArgs>(zoomBorder, nameof(zoomBorder.KeyDown)) //Assuming that ZoomBorder'name is zoomBorder
            .Subscribe(x =>
            {
                if (x.EventArgs.Key == Key.Up)
                {
                    //to do sth,such as calling command which defined in viewmodel
                }
            }).DisposeWith(disposable);
        });
```

**Reply by Shadowblitz16** [reply]
ok thankyou
